numpy/lib/arraypad.py

```diff
@@ -1165,191 +843,133 @@ def pad(array, pad_width, mode, **kwargs):
-                in enumerate(zip(pad_width, kwargs['stat_length'])):
-            newmat = _prepend_med(newmat, pad_before, chunk_before, axis)
-            newmat = _append_med(newmat, pad_after, chunk_after, axis)
-
-    elif mode == 'minimum':
-        for axis, ((pad_before, pad_after), (chunk_before, chunk_after)) \
-                in enumerate(zip(pad_width, kwargs['stat_length'])):
-            newmat = _prepend_min(newmat, pad_before, chunk_before, axis)
-            newmat = _append_min(newmat, pad_after, chunk_after, axis)
-
-    elif mode == 'reflect':
-        for axis, (pad_before, pad_after) in enumerate(pad_width):
-            if narray.shape[axis] == 0:
-                # Axes with non-zero padding cannot be empty.
-                if pad_before > 0 or pad_after > 0:
-                    raise ValueError("There aren't any elements to reflect"
-                                     " in axis {} of `array`".format(axis))
-                # Skip zero padding on empty axes.
-                continue
-
-            # Recursive padding along any axis where `pad_amt` is too large
-            # for indexing tricks. We can only safely pad the original axis
-            # length, to keep the period of the reflections consistent.
-            if ((pad_before > 0) or
-                    (pad_after > 0)) and newmat.shape[axis] == 1:
+    }
+    try:
+        unsupported_kwargs = set(kwargs) - set(allowed_kwargs[mode])
+    except KeyError:
+        raise ValueError("mode '{}' is not supported".format(mode))
+    if unsupported_kwargs:
+        raise ValueError("unsupported keyword arguments for mode '{}': {}"
+                         .format(mode, unsupported_kwargs))
+
+    stat_functions = {"maximum": np.max, "minimum": np.min,
+                      "mean": np.mean, "median": np.median}
+
+    # Create array with final shape and original values
+    # (padded area is undefined)
+    padded, old_region_sl = _pad_simple(array, pad_width)
+    # And prepare iteration over all dimensions
+    # (zipping may be more readable than using enumerate)
+    axes = range(padded.ndim)
+
+    if mode == "constant":
+        values = kwargs.get("constant_values", 0)
+        values = _as_pairs(values, padded.ndim)
+        for axis, index_pair, value_pair in zip(axes, pad_width, values):
+            roi = _view_roi(padded, old_region_sl, axis)
+            _set_pad_area(roi, axis, index_pair, value_pair)
+
+    elif mode == "empty":
+        pass  # Do nothing as padded is already prepared
+
+    elif array.size == 0:
+        # Only modes "constant" and "empty" can extend empty axes, all other
+        # modes depend on `array` not being empty
+        # -> ensure every empty axis is only "padded with 0"
+        for axis, index_pair in zip(axes, pad_width):
+            if (
+                array.shape[axis] == 0
+                and (index_pair[0] > 0 or index_pair[1] > 0)
+            ):
```

Comment:
This is strange formattering, can you start on the same line as if and just indent the second line a bit more? IIRC the `and` should go on the first line.

Reply:
Well, pep8 recommends to put binary operators on the continuation line and I took my inspiration from there. 

How about `if array.shape[axis] == 0 and any(index_pair)` instead? It fits on a single line as is perhaps more readable anyway. ;)

Reply:
Sorry about the binary operator, right of course. I doubt we use this elsewhere in numpy currently  (especially the `):` part on the new line). Can't say I like it, but I guess tastes differ and it seems to be a not-uncommon style.